### ISSUE TITLE
Ability to create index when loading table

### ISSUE THREAD

**Issue by sadikovi** [lightcopy/parquet-index#11]
Add flag `spark.sql.index.createIfNotExists` or something similar to support situation of creating index when first index scan is made for a table. Subsequent scans will use created index.

**Comment by littleJava** [comment]
hi, how can i use the 'parquet-index' in the spark query scenario, is there a tutorial ?

**Comment by sadikovi** [comment]
@littleJava This is super experimental at this point, but it works and was built for these query scenarios:). I am currently trying to increase test coverage and extend it to work with Spark 1.6.x and 2.1.x. Would appreciate your feedback on overall API and how it works.

I will put README together this weekend, here is an example and details to get you started:

#### Supported versions:
Note that it currently supports Spark 2.0.1, you can make it work for 2.0.0 by modifying `HadoopFsRelation` https://github.com/lightcopy/parquet-index/blob/master/src/main/scala/org/apache/spark/sql/execution/datasources/IndexedDataSource.scala#L63 (pass SparkSession as the first argument, now it is in currying as you can see in the code). If you need it to work for Spark 2.0.0, do this:
```scala
HadoopFsRelation(
  sparkSession = metastore.session,
  location = indexCatalog,
  partitionSchema = indexCatalog.partitionSpec().partitionColumns,
  dataSchema = indexCatalog.dataSchema().asNullable,
  bucketSpec = bucketSpec,
  fileFormat = s.fileFormat,
  options = caseInsensitiveOptions)
```

#### How to use it
In order to use it on Parquet table, you need to build index first, this issue is created to enable scenario when index does not exist, and you create it on the first read, but it is not implemented yet:)

Configuration:
- `spark.sql.index.metastore` - location of index_metastore folder, can be local or HDFS, defaults to working directory + folder name. E.g. `spark.sql.index.metastore=hdfs://host:port/my/folder`
- `spark.sql.index.parquet.bloom.enabled` - enable bloom filters for Parquet, this will be done while creating index, turned off by default. E.g. `spark.sql.index.parquet.bloom.enabled=true`

```scala
// Write sample DataFrame for example purposes
val df = spark.range(0, 1000000).withColumn("str", $"id".as("string")).withColumn("p", lit("abc"))
df.write.partitionBy("p").parquet("temp/abc.parquet")

import com.github.lightcopy.implicits._
// create index, this will create index files in index_metastore
// currently supports only long, integer, or string fields
// by default uses Parquet, I recommend enabling bloom filters
spark.index.create.mode("overwrite").indexBy($"id", $"str").table("temp/abc.parquet")

// Simple query of the table
val df = spark.index.parquet("temp/abc.parquet").filter($"str" === "1")
df.show
```

You can compare performance with this:
```scala
val df = spark.read.parquet("temp/abc.parquet").filter($"str" === "1")
df.show
```

**Comment by sadikovi** [comment]
Currently we just store metadata in index, rather than inferring schema every time and listing files. Bloom filters allow to minimise number of files to read dramatically, especially if predicate is very selective.

**Comment by sadikovi** [comment]
All API is in https://github.com/lightcopy/parquet-index/blob/master/src/main/scala/org/apache/spark/sql/DataFrameIndexManager.scala#L29

This is similar to `DataFrameReader`, you just need to import implicits from `com.github.lightcopy.implicits._`

**Comment by sadikovi** [comment]
By the way, everything should just work out of the box without any hassle (if it does not -> please, create issue). Those options are to tune for the use case.

**Comment by littleJava** [comment]
it's cool ! 
I'll try it this weekend with spark2.0.2 sql in local mode. 
ORC is too inconvenient with hive to use ~~~ 

Thank you

**Comment by sadikovi** [comment]
Agree, have similar use case:) Thank you for trying it, would appreciate feedback and any help. Will try making it generally available as a spark-package during the holidays.

**Comment by littleJava** [comment]
something unbelievable: there is no sbt-spark-package:0.2.4 in "https://dl.bintray.com/spark-packages/maven/org/spark-packages/sbt-spark-package/" , when i compile parquet-index with spark-2.0.2. 
![image](https://cloud.githubusercontent.com/assets/71588/21471716/e1531580-caf5-11e6-97f0-7743738b2706.png).

**Comment by sadikovi** [comment]
Weird, I am going to try with clean cache.

**Comment by sadikovi** [comment]
Did you try pulling repo as is and run `sbt package`?

**Comment by sadikovi** [comment]
@littleJava I am going to publish 0.1.0 on `spark-packages`, you could just pull from there, if you are still having problems with sbt.

**Comment by sadikovi** [comment]
Fixed by #17.
